docs/development/journalist_api.rst

```diff
@@ -0,0 +1,414 @@
+Journalist Interface API
+========================
+
+This document describes the endpoints for SecureDrop's Journalist Interface
+API.
+
+Versioning
+~~~~~~~~~~
+
+The API is versioned and we are currently using version 1. This is set via the
+base URL, which is:
+
+.. code:: sh
+
+  /api/v1/
+
+Content Type
+~~~~~~~~~~~~
+
+Clients should send the following headers:
+
+.. code:: sh
+
+  'Accept': 'application/json',
+  'Content-Type': 'application/json'
+
+Authentication
+~~~~~~~~~~~~~~
+
+``POST /api/v1/token`` to get a token with the username, password, and 2FA
+token in the request body:
+
+.. code:: sh
+
+  {
+  	"username": "journalist",
+  	"password": "mypasswordgoeshere",
+  	"one_time_code": "123456"
+  }
+
+Thereafter in order to authenticate to protected endpoints, send the token in
+HTTP Authorization header:
+
+.. code:: sh
+
+  Authorization: Token yourtokengoeshere
+
+This header will be checked with each API request to see if it is valid and
+not yet expired. Tokens expire after 7200 seconds (120 minutes).
+
+Endpoints
+~~~~~~~~~
+
+Root Endpoint
+-------------
+
+Does not require authentication.
+
+The root endpoint describes the available resources:
+
+.. code:: sh
+
+  GET /api/v1/
+
+Response 200 (application/json):
+
+.. code:: sh
+
+  {
+    "current_user_url": "/api/v1/user/",
+    "sources_url": "/api/v1/sources/",
+    "submissions_url": "/api/v1/submissions/"
+    "token_url": "/api/v1/token/"
+  }
+
+Sources ``[/sources]``
+----------------------
+
+Get all sources [``GET``]
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Requires authentication. Provides a list of all sources and data about them
+(such as number of documents, submissions, and their public key that replies
+should be encrypted to).
+
+.. code:: sh
+
+  GET /api/v1/sources/
+
+Response 200 (application/json):
+
+.. code:: sh
+
+  {
+      "sources": [
+          {
+              "add_star_url": "/api/v1/sources/44YGZ2R7643TXBA65ZKRO5D6QH26RJ7NVDFMQJVFSMM6WA5W3ZDXNUYKGBTEUYGFCABBUEDLQ7OKS657WKOGUHFLVDLQ75GWTOX4D4A%3D/add_star/",
```

Comment:
My understanding is that generally this pattern is used to decrease coupling between client and server (i.e. servers should be constructing URIs, not clients - clients should just follow the URIs the server provides). 

From where I sit as a developer, the reason I like REST APIs (like GitHub's) that follow this pattern, is instead of using out-of-band knowledge, I can simply query the API root endpoint, and with my knowledge of basic HTTP methods, I can merrily query around and do what I need to do without having to dig in the documentation: this makes for a very intuitive interface (aka: ["A REST API should be entered with no prior knowledge beyond the initial URI (bookmark)."](http://roy.gbiv.com/untangled/2008/rest-apis-must-be-hypertext-driven)). The SecureDrop API follows the same pattern: every endpoint can be accessed by simply following the hypertext links in responses.

Reply:
So just to clarify, I'm not blocking on this change request, but I still don't like it.

I actually think that most API interactions are done through clients/libs which are built from the API docs. Such a pattern starts with a spec as some sort of ground truth, and then the API is an implementation of that spec (which will naturally have some bugs). If someone builds a client that matches a buggy API, they will see breakage if the bug is fixed. Also I find exploratory use of APIs through URLs to be mildly scary. E.g., I don't want to `DELETE` or `PUT` without *really* knowing what's happening.